### PR TITLE
[SPARK-50460][PYTHON][CONNECT] Generalize and simplify Connect exception handling

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR introduces a generalized and streamlined approach to handling exceptions in the Spark Connect Python client.

This PR includes:

1. Replaces the large if-else block for exception type matching with a more extensible class mapping (`EXCEPTION_CLASS_MAPPING`).
2. Maintains compatibility for existing exception types while making it easier to add new ones in the future.
3. Keeps the unique handling logic for specific exceptions like `PythonException` intact.

### Why are the changes needed?

The previous exception handling logic relied on a large `if-else` block, which was not scalable or maintainable as more exception types were added.

Generalizing the exception handling mechanism reduces the complexity of the code, making it easier to extend when adding new exception types now requires minimal changes to the `EXCEPTION_CLASS_MAPPING`, avoiding duplication of logic.

### Does this PR introduce _any_ user-facing change?

No API changes, but it enhances the internal implementation of exception handling.

### How was this patch tested?

Added new UTs to verify that all existing exception handling scenarios are still supported.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by itholic** [comment]
CI passed. cc @HyukjinKwon could you take a look when you find some time?

File: python/pyspark/errors/exceptions/connect.py [hunk 1]
```diff
@@ -54,206 +54,63 @@ def convert_exception(
...
-    if "sqlState" in info.metadata:
-        sql_state = info.metadata["sqlState"]
-
-    if "errorClass" in info.metadata:
-        errorClass = info.metadata["errorClass"]
-
-    if "messageParameters" in info.metadata:
-        messageParameters = json.loads(info.metadata["messageParameters"])
-
+    classes = json.loads(info.metadata.get("classes", "[]"))
```
Comment (HyukjinKwon) [inline python/pyspark/errors/exceptions/connect.py:57]: Let's avoid json parsing if `classes` does not exist

File: python/pyspark/errors/exceptions/connect.py [hunk 2]
```diff
@@ -54,206 +54,63 @@ def convert_exception(
...
+    sql_state = info.metadata.get("sqlState")
+    error_class = info.metadata.get("errorClass")
+    message_parameters = json.loads(info.metadata.get("messageParameters", "{}"))
```
Comment (HyukjinKwon) [inline python/pyspark/errors/exceptions/connect.py:60]: ditto

File: python/pyspark/errors/exceptions/connect.py (hunk 1 above)
Comment (itholic) [reply python/pyspark/errors/exceptions/connect.py]: Applied the comments. Thanks!

**Review by HyukjinKwon — APPROVED** [review]

**Comment by itholic** [comment]
Merged to master.

**Pull request closed without merging**